pkg/controller/elasticsearch/validation/autoscaling_validation_test.go

```diff
@@ -0,0 +1,571 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package validation
+
+import (
+	"strings"
+	"testing"
+
+	commonv1 "github.com/elastic/cloud-on-k8s/pkg/apis/common/v1"
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestResourcePolicies_Validate(t *testing.T) {
+	tests := []struct {
+		name            string
+		autoscalingSpec string
+		// NodeSet name -> roles
+		nodeSets      map[string][]string
+		volumeClaims  []string
+		wantError     bool
+		expectedError string
+	}{
+		{
+			name:      "ML must be in a dedicated autoscaling policy",
+			wantError: true,
+			nodeSets:  map[string][]string{"nodeset-data-ml": {"data", "ml"}},
+			autoscalingSpec: `
+{
+	 "policies" : [{
+		  "name": "data_ml_policy",
+		  "roles": [ "data", "ml" ],
+		  "resources" : {
+			"nodeCount" : { "min" : 1 , "max" : 2 },
+			"cpu" : { "min" : 1 , "max" : 1 },
+			"memory" : { "min" : "2Gi" , "max" : "2Gi" },
+			"storage" : { "min" : "5Gi" , "max" : "10Gi" }
+		  }
+		}]
+}
+`,
+			expectedError: "ML nodes must be in a dedicated autoscaling policy",
+		},
+		{
+			name:      "ML is in a dedicated autoscaling policy",
+			wantError: false,
+			nodeSets:  map[string][]string{"nodeset-ml": {"ml"}},
+			autoscalingSpec: `
+{
+	 "policies" : [{
+		  "name": "ml",
+		  "roles": [ "ml" ],
+		  "resources" : {
+			"nodeCount" : { "min" : 1 , "max" : 2 },
+			"cpu" : { "min" : 1 , "max" : 1 },
+			"memory" : { "min" : "2Gi" , "max" : "2Gi" },
+			"storage" : { "min" : "5Gi" , "max" : "10Gi" }
+		  }
+		}]
+}
+`,
+		},
+		{
+			name:      "Happy path",
+			wantError: false,
+			nodeSets:  map[string][]string{"nodeset-data-1": {"data"}, "nodeset-data-2": {"data"}, "nodeset-ml": {"ml"}},
+			autoscalingSpec: `
+{
+	 "policies" : [{
+		  "name": "data_policy",
+		  "roles": [ "data" ],
+		  "resources" : {
+			"nodeCount" : { "min" : 1 , "max" : 2 },
+			"cpu" : { "min" : 1 , "max" : 1 },
+			"memory" : { "min" : "2Gi" , "max" : "2Gi" },
+			"storage" : { "min" : "5Gi" , "max" : "10Gi" }
+		  }
+		},
+		{
+		  "name": "ml_policy",
+		  "roles": [ "ml" ],
+		  "resources" : {
+			"nodeCount" : { "min" : 1 , "max" : 2 },
+			"cpu" : { "min" : 1 , "max" : 1 },
+			"memory" : { "min" : "2Gi" , "max" : "2Gi" },
+			"storage" : { "min" : "5Gi" , "max" : "10Gi" }
+		  }
+		}]
+}
+`,
+		},
+		{
+			name:          "Autoscaling policy with no NodeSet",
+			wantError:     true,
+			expectedError: "Invalid value: []string{\"ml\"}: roles must be set in at least one nodeSet",
```

Comment:
```suggestion
			expectedError: "Invalid value: []string{\"ml\"}: roles must be used in at least one nodeSet",
```

is maybe a tiny bit clearer.